proofs/clenvtac.ml

```diff
@@ -84,16 +84,255 @@ let clenv_refine ?(with_evars=false) ?(with_classes=true) clenv =
+    let v = EConstr.of_constr (nf_betaiota env (EConstr.Unsafe.to_constr (clenv_val clenv))) in
+    (* This renaming hack should really stop at 8.6 *)
+    if Flags.version_less_or_equal Flags.Current then
+      rename_term env sigma v
+    else sigma, v
+  in
+  let reduce_goal gl =
+    let sigma = Proofview.Goal.sigma gl in
+    let concl = Proofview.Goal.concl gl in
+    let glev = Proofview.Goal.goal gl in
+    (* For compatibility: beta iota reduction *)
+    let concl = Reductionops.clos_norm_flags CClosure.betaiota env sigma concl in
+    let evi = Evd.find sigma glev in
+    let sigma = Evd.add sigma glev { evi with evar_concl = concl } in
+    let sigma = Typeclasses.make_unresolvables (fun ev -> Evar.equal ev glev) sigma in
+    Proofview.Unsafe.tclEVARS sigma
+  in
+  Proofview.Unsafe.tclEVARS sigma <*>
+    clenv_check_dep_holes with_evars env sigma ?origsigma clenv >>= (fun deps ->
+  (Refine.refine ~typecheck:false run) <*>
+  (if shelve_subgoals then shelve_goals deps else tclUNIT ()) <*>
+    Proofview.Goal.enter reduce_goal)
+  end
+
+open Unification
+
+open Proofview.Notations
+
+let clenv_refine_no_check
+      ?(with_evars=false) ?(with_classes=true) ?(shelve_subgoals=true)
+      ?(flags=dft ()) ?origsigma clenv =
+  let flags = flags_of flags in
+  Proofview.tclEVARMAP >>= fun sigma ->
+  clenv_refine_gen ~with_evars ~with_classes ~shelve_subgoals ?origsigma
+                   flags (sigma, clenv)
+
+let clenv_refine2 ?(with_evars=false) ?(with_classes=true) ?(shelve_subgoals=true)
+                  ?(flags=dft ()) ?origsigma clenv =
+  let flags = flags_of flags in
+  let tac = clenv_unify_concl flags clenv in
+  Ftactic.run tac
+   (clenv_refine_gen ~with_evars ~with_classes ~shelve_subgoals ?origsigma flags)
+
+(** The dependent holes turned into subgoals are
+    evars of the clause which are dependent in other hypotheses of the clause,
+    whether or not they appear in the instantiated conclusion.
+ *)
+
+let clenv_refine_bindings
+    ?(with_evars=false) ?(with_classes=true) ?(shelve_subgoals=true)
+    ?(flags=dft ()) ~hyps_only ~delay_bindings b ?origsigma clenv =
+  let open Proofview in
+  let flags = flags_of flags in
+  Proofview.Goal.enter begin fun gl ->
+    let env = Proofview.Goal.env gl in
+    let sigma = Proofview.Goal.sigma gl in
+    let sigma, clenv, bindings =
+      if delay_bindings then
+        sigma, clenv, Some (None, false, b)
+      else
+        try let sigma, clenv = Clenv.solve_evar_clause env sigma ~hyps_only clenv b in
+            sigma, clenv, None
+        with e -> sigma, clenv, Some (Some e, hyps_only, b)
+    in
+    let tac = clenv_unify_concl flags clenv in
+    (Unsafe.tclEVARS sigma) <*>
+    (Ftactic.run tac
+      (fun (sigma, clenv) ->
+        try let sigma, clenv =
+          match bindings with
+          | Some (exn, hyps_only, b) ->
+             (* Hack to make [exists 0] on [Σ x : nat, True] work, we
```

Comment:
@mattam82 Why is this `delay_bindings` hack needed? It makes `split` fail on `@sigT True (fun _ => True)`. Removing it gives me a universe inconsistency in `JMeq` (on an `exists`), but not necessarily for good reasons.

Reply:
Because your `split` shouldn’t work? I tried to maintain the exact same semantics as before here, for compatibility, otherwise it was breaking loads of code. I’m not saying this is not a hack :)

Reply:
Well, it seems used through `dtauto` in `test-suite/bugs/opened/bug_6393.v`, which now fails.

Reply:
And I confirm `split` does work on `@sigT True (fun _ => True)` on master.

Reply:
Ah, interesting, it certainly got fixed in the meantime. Can you remind me how split calls apply ?

Reply:
So that's `constructor` in fact.
On master: https://github.com/coq/coq/blob/831639deec0ce88fca4ede4756815cf434088ac3/tactics/tactics.ml#L2203
on this branch: 
https://github.com/mattam82/coq/blob/98326adaf66b7876bae7f4fb43e7530841377715/tactics/tactics.ml#L2265